internal/route/request_executor.go

```diff
@@ -50,6 +49,89 @@ func (r *RequestExecutor) routeToConfig(
 		httpReq.Header.Set("Authorization", "Basic "+encodedCredentials)
 	}
 
+	var (
+		respBody jsonrpc.ResponseBody
+		resp     *http.Response
+	)
+
+	_, isSingleRequestBody := requestBody.(*jsonrpc.SingleRequestBody)
+
+	if r.cache != nil && r.cache.ShouldCacheMethod(requestBody.GetMethod()) && isSingleRequestBody {
+		// We can not send the same http.Request twice.
+		// In cache of a cache miss and the request to origin fails.
+		respBody, resp, err = r.retrieveOrCacheRequest(cloneRequest(httpReq), requestBody, configToRoute)
+		if err != nil {
+			r.logger.Warn("caching error", zap.Error(err), zap.Any("request", requestBody))
+		}
+	}
+
+	// If resp is nil, then either caching is disabled or caching failed.
+	// We want to send the request to the origin in this case.
+	if resp == nil {
+		respBody, resp, err = r.getResponseBody(httpReq, requestBody, configToRoute)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	return respBody, resp, nil
+}
+
+func (r *RequestExecutor) retrieveOrCacheRequest(httpReq *http.Request, requestBody jsonrpc.RequestBody, configToRoute *config.UpstreamConfig) (jsonrpc.ResponseBody, *http.Response, error) {
```

Comment:
Should we move most of the logic of this function into RPCCache? An interface could be the method name and params.

Working with redis directly feels like a pretty low level of abstraction for the RequestExecutor.

Reply:
Took a little work, but I managed to move redis interactions into the RPCCache. It does rely on some more closures and passing functions, so readability dropped a little bit.